compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp

```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
+            /*allowInvalidState=*/true);
+        if (!maybeAccessor) {
+          return WalkResult::advance();
+        }
+        if (maybeAccessor->isValidState()) {
+          auto accessorMap = maybeAccessor->getAssumedMultiSet();
+          // If the map is empty, this operation has no hazards for accessor
+          // motion. Because this is true for most operations (i.e. anything
+          // regionless and not specifically marked with a relevant side effect)
+          // we elide the map in such cases.
+          if (!accessorMap.empty()) {
+            accessorCounts[op] = accessorMap;
+            globalAccessors.insert(op);
+          }
+        } else {
+          globalAccessors.insert(op);
+        }
+        if (auto store = dyn_cast<IREE::Util::GlobalStoreOpInterface>(op)) {
+          bucketedAccessors[store.getGlobalAttr().getAttr()].insert(store);
+        } else if (auto load =
+                       dyn_cast<IREE::Util::GlobalLoadOpInterface>(op)) {
+          bucketedAccessors[load.getGlobalAttr().getAttr()].insert(load);
+        } else if (auto loop = dyn_cast<LoopLikeOpInterface>(op)) {
+          compositeAccessors.push_back(loop);
+        }
+        return WalkResult::advance();
+      });
+    }
+    return result;
+  }
+
+  // Process the given load by moving up to the nearest blocking operation and
+  // performing RAR or RAW forwarding if the blocking operation is a load/store.
+  bool processGlobalLoad(IREE::Util::GlobalLoadOpInterface load);
+  // Process the given store by moving down to the nearest blocking operation
+  // and performing WAW forwarding if the blocking operation is a store.
+  bool processGlobalStore(IREE::Util::GlobalStoreOpInterface store);
+  // Perform LICM style hoisting of single globals accessors.
+  bool processLoopLikeOp(LoopLikeOpInterface loop);
+  // Process the given loop, hoisting redundant accessors. On success this
+  // returns the newly created loop and replaces the accessor map entry for
+  // the original loop with the new one.
+  Operation *hoistRedundantAccessors(LoopLikeOpInterface loop);
+  // Process all composite accessors (e.g. scf.for).
+  bool processCompositeAccessors();
+  // Process all of loads/stores for the given global.
+  bool processDirectAccessors(StringAttr globalName);
+  // Process all accessors by bucket in |bucketedAccessors|.
+  bool processGlobalAccessors();
+
+private:
+  // Move an operation right after the nearest blocking operation above |op|
+  // within its block. An operation is blocking if it *may* access the global
+  // referenced by |globalName|.
+  void moveOpUpInBlock(Operation *op, StringAttr globalName);
+  // Move an operation right before the nearest blocking operation below |op|
+  // within its block. An operation is blocking if it *may* access the global
+  // referenced by |globalName|.
+  void moveOpDownInBlock(Operation *op, StringAttr globalName);
+
+  StringAttr getGlobalName(Operation *op) {
```

Comment:
if you change to using GlobalInfo then you can just ask the globalInfo.op for its name

Reply:
The propagation is rooted on `util.global.load`/`util.global.store`, and this is just supposed to be a helper for getting the name of the global from the attribute already on the op. I am already maintaining my own cache of Operation* -> Map<GlobalName, AccessCount> because I want to know, given an operation, does it access a particular global. This is what the analysis constructs and is easier to reuse directly. The `initializeOperation` that the solver internally calls back to when constructing a new element only gives me the operation and the solver itself, so it's much easier to just record the global name already tied to the `util.global.load`/`util.global.store` than to try to go and figure out which GlobalInfo it's tied to.